### PR TITLE
🎉 Update algolia index when publishing/updating/unpublishing gdoc posts

### PR DESCRIPTION
**Pull request by ikesau**
Resolves https://github.com/owid/owid-grapher/issues/2852

Adds a new step to the dreaded gdocs PUT route ☠️ 

**If publishing/updating:**
Update the pages index with new records

**If deleting/unpublishing:**
Delete the records from the pages index

I had to move stuff out of `indexToAlgolia.tsx` so that I could import a function without the implicit call to `indexToAlgola()` being evaluated.

Also added some comments about Fragments and https://github.com/owid/owid-grapher/issues/3426 as I encountered that annoying little gotcha.

## To test:
Set `ALGOLIA_INDEXING=true`, `ALGOLIA_ID`, `ALGOLIA_SEARCH_KEY`, `ALGOLIA_SECRET_KEY` with values from the owid-staging application on Algolia and set  in your local `.env` file

Publish a test post with at least 1500 characters (for chunking) and verify that it shows up in the owid-staging [pages index](https://dashboard.algolia.com/apps/74GKBOIDJQ/explorer/browse/pages)

Update the post to have a short amount of content and republish it. Confirm that the content is updated in the index and that records from the larger version of the post are deleted.

Change the slug of the post and republish it again. Confirm that the records are updated correctly.

Unpublish the post. Confirm that the records are deleted from the index.

Publish the post again, confirm that new records are created, and then delete the post (without unpublishing first). Confirm that the records are deleted.

Confirm that fragments don't get indexed but topic pages, linear topic pages, about pages, and articles do.

### CODE REVIEW

**Review by marcelgerber — APPROVED** [review]
Nice!
It's a bummer that the gdocs PUT route is so full now; would be nice if we can factor some things out there at some point.

File: baker/algolia/algoliaUtils.tsx [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+            url: gdoc.slug,
+        },
+    }
+    const records = generateGdocRecords([gdoc], pageviewsForGdoc)
+
+    const existingRecordsForPost: ObjectWithObjectID[] = []
+    await index.browseObjects({
+        // IMPORTANT: requires `slug` to be set in the index's attributesForFaceting
+        filters: `slug:${indexedSlug}`,
+        attributesToRetrieve: ["ObjectID"],
```
[the remark this answers is not in the document]
Comment (ikesau) [reply baker/algolia/algoliaUtils.tsx:278]: It's actually returned by default and doesn't need to be specified. I was just adding it to help make the code more understandable, so good catch! 😬

File: adminSiteServer/apiRouter.ts [hunk 2]
```diff
@@ -2380,12 +2387,17 @@ putRouteWithRWTransaction(apiRouter, "/gdocs/:id", async (req, res, trx) => {
             nextJson.slug
         )
     } else if (checkFullDeployFallback(prevJson, nextJson, hasChanges)) {
```
Comment (marcelgerber) [inline adminSiteServer/apiRouter.ts:2389]: What's a bit weird is that lightning deploys don't trigger a `GdocPublishingAction.Updating` action.
I don't know the numbers, but they might be the more common case than fallback deploys, no?

Comment (ikesau) [reply adminSiteServer/apiRouter.ts]: True! I've fixed lightning deploys: a function property was missing from the properties to exclude list and so `isEqual(prev, next)` was always returning `false` - not great that it's so easy to miss this but I think it was mostly due to the large gdocs model refactor and should be easier to notice again now that it's working.

I also extracted functions from the PUT route (probably easier to review in an editor than the git diff)  most notably `indexAndBakeGdocIfNeccesary` which handles all four possible publishing action cases more explicitly, now, and should make it easier to update.

File: adminSiteServer/apiRouter.ts [hunk 3]
```diff
@@ -2287,67 +2293,75 @@ getRouteNonIdempotentWithRWTransaction(
...
 )
 
+/**
+ * Handles all four `GdocPublishingAction` cases
+ * - Saving (no action)
+ * - Publishing (index and bake)
+ * - Updating (index and bake (potentially via lightning deploy))
+ * - Unpublishing (remove from index and bake)
+ */
+async function indexAndBakeGdocIfNeccesary(
```
Comment (marcelgerber) [inline adminSiteServer/apiRouter.ts:2298]: Maybe it makes sense to move this one to `gdocsDeploy`, too?

**Review by marcelgerber — APPROVED** [review]
Nice refactor, that's making this much better. Thank you!

File: adminSiteServer/apiRouter.ts (hunk 3 above)
Comment (ikesau) [reply adminSiteServer/apiRouter.ts:2298]: Yeahhh I had the same desire but those functions are in the `adminSiteClient` folder, and I didn't want to make a new file just for one function so I kept it close to usage instead.

**Pull request merged**